webrtc/udp_connection.nim

```diff
@@ -0,0 +1,89 @@
+# Nim-WebRTC
+# Copyright (c) 2024 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import errors
+import chronos, chronicles
+
+logScope:
+  topics = "webrtc udp"
+
+# UdpConn is a small wrapper of the chronos DatagramTransport.
+# It's the simplest solution we found to store the message and
+# the remote address used by the underlying protocols (dtls/sctp etc...)
+
+type
+  WebRtcUdpError = object of WebRtcError
+
+  UdpPacketInfo* = tuple
+    message: seq[byte]
+    raddr: TransportAddress
+
+  UdpConn* = ref object
+    laddr*: TransportAddress
+    udp: DatagramTransport
+    dataRecv: AsyncQueue[UdpPacketInfo]
+    closed: bool
+
+proc init*(self: UdpConn, laddr: TransportAddress) =
+  ## Initialize an Udp Connection
+  ##
+  self.laddr = laddr
+  self.closed = false
+
+  proc onReceive(
+      udp: DatagramTransport,
+      raddr: TransportAddress
+    ) {.async: (raises: []), gcsafe.} =
+    # On receive Udp message callback, store the
+    # message with the corresponding remote address
+    try:
+      trace "UDP onReceive"
+      let msg = udp.getMessage()
+      self.dataRecv.addLastNoWait((msg, raddr))
+    except CatchableError as exc:
+      raiseAssert(exc.msg)
+
+  self.dataRecv = newAsyncQueue[UdpPacketInfo]()
+  self.udp = newDatagramTransport(onReceive, local = laddr)
+
+proc close*(self: UdpConn) =
+  ## Close an Udp Connection
+  ##
+  if self.closed:
+    debug "Trying to close an already closed UdpConn"
+    return
+  self.closed = true
+  self.udp.close()
+
+proc write*(
+    self: UdpConn,
+    raddr: TransportAddress,
+    msg: seq[byte]
+  ) {.async: (raises: [CancelledError, WebRtcUdpError]).} =
+  ## Write a message on Udp to a remote address `raddr`
+  ##
+  if self.closed:
+    debug "Try to write on an already closed UdpConn"
+    return
+  trace "UDP write", msg
+  try:
+    await self.udp.sendTo(raddr, msg)
+  except TransportError as exc:
+    raise newException(WebRtcUdpError, exc.msg, exc)
+  except CancelledError as exc:
```

Comment:
If we catch and reraise without doing anything, I think we can just remove this.

Reply:
Hum... You're right, it's a bad habit, catching/reraising is only useful when we catch a `CatchableError` before.

Reply:
done